npm/pkg/dataplane/ipsets/dirtycache.go

```diff
@@ -0,0 +1,264 @@
+	toUpdateCache  map[string]*memberDiff
+	toDestroyCache map[string]*memberDiff
+}
+
+func newDirtyCache() *dirtyCache {
+	dc := &dirtyCache{}
+	dc.reset()
+	return dc
+}
+
+func (dc *dirtyCache) reset() {
+	dc.toCreateCache = make(map[string]*memberDiff)
+	dc.toUpdateCache = make(map[string]*memberDiff)
+	dc.toDestroyCache = make(map[string]*memberDiff)
+}
+
+func (dc *dirtyCache) resetAddOrUpdateCache() {
+	dc.toCreateCache = make(map[string]*memberDiff)
+	dc.toUpdateCache = make(map[string]*memberDiff)
+}
+
+func (dc *dirtyCache) create(set *IPSet) {
+	// error checking
+	if _, ok := dc.toCreateCache[set.Name]; ok {
+		msg := fmt.Sprintf("create: set %s should not already be in the toCreateCache", set.Name)
+		klog.Error(msg)
+		metrics.SendErrorLogAndMetric(util.IpsmID, msg)
+		return
+	}
+	if _, ok := dc.toUpdateCache[set.Name]; ok {
+		msg := fmt.Sprintf("create: set %s should not be in the toUpdateCache", set.Name)
+		klog.Error(msg)
+		metrics.SendErrorLogAndMetric(util.IpsmID, msg)
+		return
+	}
+
+	diff, ok := dc.toDestroyCache[set.Name]
+	if ok {
+		// transfer from toDestroyCache to toUpdateCache and maintain member diff
+		dc.toUpdateCache[set.Name] = diff
+		delete(dc.toDestroyCache, set.Name)
+	} else {
+		// put in the toCreateCache
+		dc.toCreateCache[set.Name] = diffOnCreate(set)
+	}
+}
+
+// could optimize Linux to remove from toUpdateCache if there were no member diff afterwards,
+// but leaving as is prevents difference between OS caches
+func (dc *dirtyCache) addMember(set *IPSet, member string) {
+	// error checking
+	if dc.isSetToDelete(set.Name) {
+		msg := fmt.Sprintf("addMember: set %s should not be in the toDestroyCache", set.Name)
+		klog.Error(msg)
+		metrics.SendErrorLogAndMetric(util.IpsmID, msg)
+		return
+	}
+
+	diff, ok := dc.toCreateCache[set.Name]
+	if !ok {
+		diff, ok = dc.toUpdateCache[set.Name]
+		if !ok {
+			diff = newMemberDiff()
+			dc.toUpdateCache[set.Name] = diff
+		}
+	}
+	diff.addMember(member)
+}
+
+// could optimize Linux to remove from toUpdateCache if there were no member diff afterwards,
+// but leaving as is prevents difference between OS caches
+func (dc *dirtyCache) deleteMember(set *IPSet, member string) {
+	// error checking
+	if dc.isSetToDelete(set.Name) {
+		msg := fmt.Sprintf("deleteMember: set %s should not be in the toDestroyCache", set.Name)
+		klog.Error(msg)
+		metrics.SendErrorLogAndMetric(util.IpsmID, msg)
+		return
+	}
+
+	diff, ok := dc.toCreateCache[set.Name]
+	if ok {
+		// don't mark a member to be deleted if it never existed in the kernel
+		diff.deleteMemberIfExists(member)
+	} else {
+		diff, ok = dc.toUpdateCache[set.Name]
+		if !ok {
+			diff = newMemberDiff()
+			dc.toUpdateCache[set.Name] = diff
+		}
+		diff.deleteMember(member)
+	}
+}
+
+func (dc *dirtyCache) destroy(set *IPSet) {
+	// error checking
+	if dc.isSetToDelete(set.Name) {
+		msg := fmt.Sprintf("destroy: set %s should not already be in the toDestroyCache", set.Name)
+		klog.Error(msg)
```

Comment:
same here can be a warning, do not need an error here, and also, are these warnings required ? Do these indicate problems elsewhere in DP/setMgr code?

Reply:
the error checks make sure that ipsetmanager is following the assumptions of the dirty cache. Assumptions are in the comments of the dirtyCacheInterface

Reply:
talked offline. Safe to make multiple destroy calls in a row